packages/compose/src/hooks/use-resize-observer/index.ts

```diff
@@ -0,0 +1,186 @@
+		 */
+		element: T
+	) => void,
+	/**
+	 * Options that, with the exception of `targetElement`, will be passed to
+	 * `ResizeObserver.observe` when called internally.
+	 *
+	 * Updating them will not cause the observer to be re-created, and they will
+	 * only take effect if a new element is observed.
+	 */
+	options?: ObserveElementSizeOptions< T >
+): ( element?: T | null | undefined ) => void;
+
+/**
+ * **This is a legacy API and should not be used.**
+ *
+ * @deprecated Use the other `useResizeObserver` API instead: `const ref = useResizeObserver( ( entries ) => { ... } )`.
+ *
+ * Hook which allows to listen to the resize event of any target element when it changes size.
+ * _Note: `useResizeObserver` will report `null` sizes until after first render.
+ *
+ * @example
+ *
+ * ```js
+ * const App = () => {
+ * 	const [ resizeListener, sizes ] = useResizeObserver();
+ *
+ * 	return (
+ * 		<div>
+ * 			{ resizeListener }
+ * 			Your content here
+ * 		</div>
+ * 	);
+ * };
+ * ```
+ */
+export default function useResizeObserver(): [ ReactElement, ObservedSize ];
+
+export default function useResizeObserver< T extends HTMLElement >(
+	onUpdate?: (
+		resizeObserverEntries: ResizeObserverEntry[],
+		element: T
+	) => void,
+	options: ObserveElementSizeOptions< T > = {}
+):
+	| ( ( element?: T | null | undefined ) => void )
+	| [ ReactElement, ObservedSize ] {
+	return onUpdate
+		? _useResizeObserver( onUpdate, options )
+		: _useLegacyResizeObserver();
+}
+
+function _useResizeObserver< T extends HTMLElement >(
+	onUpdate: (
+		resizeObserverEntries: ResizeObserverEntry[],
+		element: T
+	) => void,
+	{
+		targetElement,
+		...resizeObserverOptions
+	}: ObserveElementSizeOptions< T > = {}
+): ( element?: T | null | undefined ) => void {
+	const onUpdateEvent = useEvent( onUpdate );
+
+	const observedElementRef = useRef< T | null >();
+	const resizeObserverRef = useRef< ResizeObserver >();
+	const setElement = useEvent( ( element?: T | null | undefined ) => {
+		if ( element === observedElementRef.current ) {
+			return;
+		}
+		observedElementRef.current = element;
+
+		// Set up `ResizeObserver`.
+		if ( ! resizeObserverRef.current ) {
+			resizeObserverRef.current = new ResizeObserver( ( entries ) => {
+				if ( observedElementRef.current ) {
+					onUpdateEvent( entries, observedElementRef.current );
+				}
+			} );
+		}
+		const { current: resizeObserver } = resizeObserverRef;
+
+		// Unobserve previous element.
+		if ( observedElementRef.current ) {
+			resizeObserver.unobserve( observedElementRef.current );
+		}
+
+		// Observe new element.
+		if ( element ) {
+			resizeObserver.observe( element, resizeObserverOptions );
+		}
+	} );
+
+	// Handle `targetElement` option.
+	const isArgumentRef = useRef( Boolean( targetElement ) );
+	useLayoutEffect( () => {
+		if ( targetElement || isArgumentRef.current ) {
+			isArgumentRef.current = true;
+			setElement( targetElement );
+			return setElement;
```

Comment:
The main call can already be `undefined` if the passed option is, so there's no point in trying to remove `undefined` from the type.

Reply:
Kept the `undefined` in the union since the returned setter is meant to be used in more ways than just a callback ref, see https://github.com/WordPress/gutenberg/pull/64943#issuecomment-2337931758

Resolving, but we can always do a follow-up!